### PR TITLE
VS-156: Support Strong Named 2.28 Driver

### PR DESCRIPTION
**Pull request by Ravi-Raghavan**

### CODE REVIEW

File: tests/MongoDB.Analyzer.Tests/Infrastructure/DiagnosticsAnalyzer.cs [hunk 1]
```diff
@@ -73,8 +78,79 @@ internal static class DiagnosticsAnalyzer
         var analyzerOptions = new AnalyzerOptions(ImmutableArray.Create<AdditionalText>(new AdditionalTextAnalyzerSettings(settings)));
 
         var compilationWithAnalyzers = compilation.WithAnalyzers(ImmutableArray.Create<DiagnosticAnalyzer>(mongodbAnalyzer), analyzerOptions);
+
+        var test = PrintAssemblyVersion(compilation, "MongoDB.Bson");
```
Comment (BorisDog) [inline tests/MongoDB.Analyzer.Tests/Infrastructure/DiagnosticsAnalyzer.cs:82]: Cleanup

File: tests/MongoDB.Analyzer.Tests/Infrastructure/DiagnosticsAnalyzer.cs [hunk 2]
```diff
@@ -46,20 +53,18 @@ internal static class DiagnosticsAnalyzer
...
         var testCodeText = File.ReadAllText(testCodeFilename);
         var testCodeSyntaxTree = CSharpSyntaxTree.ParseText(testCodeText);
-        var compilationOptions = new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary);
 
-        var compilation = CSharpCompilation.Create(
-            "TestAssembly",
-            new[] { testCodeSyntaxTree },
-            metadataReferences,
-            compilationOptions);
+        var publicKey = new byte[160] { 0, 36, 0, 0, 4, 128, 0, 0, 148, 0, 0, 0, 6, 2, 0, 0, 0, 36, 0, 0, 82, 83, 65, 49, 0, 4, 0, 0, 1, 0, 1, 0, 53, 40, 127, 13, 56, 131, 192, 160, 117, 200, 142, 12, 218, 60, 233, 59, 98, 16, 3, 236, 189, 94, 146, 13, 74, 140, 114, 56, 86, 79, 77, 47, 79, 104, 17, 106, 202, 40, 201, 178, 19, 65, 220, 58, 135, 118, 121, 193, 69, 86, 25, 43, 43, 47, 95, 226, 193, 29, 98, 78, 8, 148, 211, 8, 255, 123, 148, 191, 111, 215, 42, 239, 27, 65, 1, 127, 254, 37, 114, 233, 144, 25, 209, 198, 25, 99, 230, 140, 208, 237, 103, 115, 74, 66, 203, 51, 59, 128, 142, 56, 103, 203, 230, 49, 147, 114, 20, 227, 46, 64, 159, 177, 250, 98, 253, 182, 157, 73, 76, 37, 48, 230, 74, 64, 228, 23, 214, 238 };
```
Comment (BorisDog) [inline tests/MongoDB.Analyzer.Tests/Infrastructure/DiagnosticsAnalyzer.cs:64]: Probably not needed

File: src/MongoDB.Analyzer/Core/Utilities/AnalysisCodeGeneratorUtilities.cs [hunk 3]
```diff
@@ -18,31 +18,49 @@ namespace MongoDB.Analyzer.Core.Utilities;
...
     {
-        var compilation = CSharpCompilation.Create(
-            AnalysisAssemblyName,
-            syntaxTrees,
-            referencesContainer.References,
-            new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary));
-
-        using var memoryStream = new MemoryStream();
-        var emitResult = compilation.Emit(memoryStream);
+        var publicKey = new byte[160] { 0, 36, 0, 0, 4, 128, 0, 0, 148, 0, 0, 0, 6, 2, 0, 0, 0, 36, 0, 0, 82, 83, 65, 49, 0, 4, 0, 0, 1, 0, 1, 0, 53, 40, 127, 13, 56, 131, 192, 160, 117, 200, 142, 12, 218, 60, 233, 59, 98, 16, 3, 236, 189, 94, 146, 13, 74, 140, 114, 56, 86, 79, 77, 47, 79, 104, 17, 106, 202, 40, 201, 178, 19, 65, 220, 58, 135, 118, 121, 193, 69, 86, 25, 43, 43, 47, 95, 226, 193, 29, 98, 78, 8, 148, 211, 8, 255, 123, 148, 191, 111, 215, 42, 239, 27, 65, 1, 127, 254, 37, 114, 233, 144, 25, 209, 198, 25, 99, 230, 140, 208, 237, 103, 115, 74, 66, 203, 51, 59, 128, 142, 56, 103, 203, 230, 49, 147, 114, 20, 227, 46, 64, 159, 177, 250, 98, 253, 182, 157, 73, 76, 37, 48, 230, 74, 64, 228, 23, 214, 238 };
```
Comment (BorisDog) [inline src/MongoDB.Analyzer/Core/Utilities/AnalysisCodeGeneratorUtilities.cs:26]: Try to inspect ReferencesContainer.Version to decide with name to use

File: evergreen/run-tests.sh [hunk 4]
```diff
@@ -6,6 +6,8 @@ echo Testing with driver: "${DRIVER_VERSION}" "${TARGET_FRAMEWORK}"
 
 dotnet clean "./MongoDB.Analyzer.sln"
 
-dotnet build "./tests/MongoDB.Analyzer.Tests.Common.ClassLibrary" -f netstandard2.0 -c Debug
+dotnet build "./tests/MongoDB.Analyzer.Tests.Common.ClassLibrary" -f netstandard2.0 -c DRIVER_2_27_OR_LOWER
+
```
Comment (BorisDog) [inline evergreen/run-tests.sh:10]: Remove empty line (group together the ClassLibrary builds)

**Review by BorisDog — CHANGES_REQUESTED** [review]

File: tests/MongoDB.Analyzer.Tests/Infrastructure/DiagnosticsAnalyzer.cs (hunk 1 above)
Comment (Ravi-Raghavan) [reply tests/MongoDB.Analyzer.Tests/Infrastructure/DiagnosticsAnalyzer.cs:82]: Done

File: tests/MongoDB.Analyzer.Tests/Infrastructure/DiagnosticsAnalyzer.cs (hunk 2 above)
Comment (Ravi-Raghavan) [reply tests/MongoDB.Analyzer.Tests/Infrastructure/DiagnosticsAnalyzer.cs:64]: Done

File: evergreen/run-tests.sh (hunk 4 above)
Comment (Ravi-Raghavan) [reply evergreen/run-tests.sh:10]: Done

File: src/MongoDB.Analyzer/Core/Utilities/AnalysisCodeGeneratorUtilities.cs (hunk 3 above)
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/Core/Utilities/AnalysisCodeGeneratorUtilities.cs:26]: Done

File: src/MongoDB.Analyzer/Core/Utilities/AnalysisCodeGeneratorUtilities.cs [hunk 5]
```diff
@@ -13,20 +13,23 @@
...
+using NuGet.Versioning;
 
 namespace MongoDB.Analyzer.Core.Utilities;
 
 internal static class AnalysisCodeGeneratorUtilities
 {
-    private const string AnalysisAssemblyName = "DynamicProxyGenAssembly2";
+    private const string AnalysisAssemblyNameDRIVER_2_27_OR_Lower = "DynamicProxyGenAssembly2";
+    private const string AnalysisAssemblyNameDRIVER_2_28_OR_Greater = "MongoDB.Analyzer.MQLGenerator";
+    private const string Driver_V2_28_OrGreater = "[2.28.0, )";
```
Comment (BorisDog) [inline src/MongoDB.Analyzer/Core/Utilities/AnalysisCodeGeneratorUtilities.cs:24]: Let's have consistent naming style:
 28_OR_GREATER for ifdef directives and configurations. And OrLower/OrGreater for code variables.

File: src/MongoDB.Analyzer/Core/Utilities/AnalysisCodeGeneratorUtilities.cs [hunk 6]
```diff
@@ -56,5 +59,7 @@ analysisType switch
             AnalysisType.Poco => JsonSyntaxElements.Poco.JsonGeneratorFullName,
             _ => throw new ArgumentOutOfRangeException(nameof(analysisType), analysisType, "Unsupported analysis type")
         };
+
+    private static bool IsDriverVersion_2_28_Or_Greater(string driverVersion) => VersionRange.Parse(Driver_V2_28_OrGreater).Satisfies(NuGetVersion.Parse(driverVersion));
```
Comment (BorisDog) [inline src/MongoDB.Analyzer/Core/Utilities/AnalysisCodeGeneratorUtilities.cs:63]: Consider having `readonly Version s_driverVersion_2_28`, and comparing 
```
referencesContainer.Version >=
 s_driverVersion_2_28
```
We don't actually need range here.

File: tests/MongoDB.Analyzer.Tests/Infrastructure/DiagnosticsAnalyzer.cs [hunk 7]
```diff
@@ -46,20 +48,16 @@ internal static class DiagnosticsAnalyzer
...
         var testCodeText = File.ReadAllText(testCodeFilename);
         var testCodeSyntaxTree = CSharpSyntaxTree.ParseText(testCodeText);
-        var compilationOptions = new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary);
 
-        var compilation = CSharpCompilation.Create(
-            "TestAssembly",
-            new[] { testCodeSyntaxTree },
-            metadataReferences,
-            compilationOptions);
+        var compilationOptions = new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary);
```
Comment (BorisDog) [inline tests/MongoDB.Analyzer.Tests/Infrastructure/DiagnosticsAnalyzer.cs:59]: If there are no changes here, revert to keep the styling?

File: src/MongoDB.Analyzer/MongoDB.Analyzer.csproj [hunk 8]
```diff
@@ -31,6 +31,7 @@
     <PackageReference Include="Microsoft.CSharp" Version="4.7.0" />
     <PackageReference Include="Newtonsoft.Json" Version="12.0.3" />
     <PackageReference Include="System.Runtime.Loader" Version="4.3.0" />
+    <PackageReference Include="NuGet.Versioning" Version="5.10.0" />
```
Comment (BorisDog) [inline src/MongoDB.Analyzer/MongoDB.Analyzer.csproj:34]: Not needed of sticking with Version.

File: src/MongoDB.Analyzer/Core/Utilities/AnalysisCodeGeneratorUtilities.cs [hunk 9]
```diff
@@ -13,20 +13,23 @@
...
 
     public static Type CompileAndGetGeneratorType(AnalysisType analysisType, MongoAnalysisContext context, ReferencesContainer referencesContainer, IEnumerable<SyntaxTree> syntaxTrees)
     {
-        var compilation = CSharpCompilation.Create(
-            AnalysisAssemblyName,
-            syntaxTrees,
-            referencesContainer.References,
-            new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary));
+        var publicKey = new byte[160] { 0, 36, 0, 0, 4, 128, 0, 0, 148, 0, 0, 0, 6, 2, 0, 0, 0, 36, 0, 0, 82, 83, 65, 49, 0, 4, 0, 0, 1, 0, 1, 0, 53, 40, 127, 13, 56, 131, 192, 160, 117, 200, 142, 12, 218, 60, 233, 59, 98, 16, 3, 236, 189, 94, 146, 13, 74, 140, 114, 56, 86, 79, 77, 47, 79, 104, 17, 106, 202, 40, 201, 178, 19, 65, 220, 58, 135, 118, 121, 193, 69, 86, 25, 43, 43, 47, 95, 226, 193, 29, 98, 78, 8, 148, 211, 8, 255, 123, 148, 191, 111, 215, 42, 239, 27, 65, 1, 127, 254, 37, 114, 233, 144, 25, 209, 198, 25, 99, 230, 140, 208, 237, 103, 115, 74, 66, 203, 51, 59, 128, 142, 56, 103, 203, 230, 49, 147, 114, 20, 227, 46, 64, 159, 177, 250, 98, 253, 182, 157, 73, 76, 37, 48, 230, 74, 64, 228, 23, 214, 238 };
```
Comment (BorisDog) [inline src/MongoDB.Analyzer/Core/Utilities/AnalysisCodeGeneratorUtilities.cs:28]: Make the key static and readonly

**Review by BorisDog — CHANGES_REQUESTED** [review]
Few more changes requested.

File: src/MongoDB.Analyzer/Core/Utilities/AnalysisCodeGeneratorUtilities.cs (hunk 5 above)
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/Core/Utilities/AnalysisCodeGeneratorUtilities.cs:24]: Done

File: src/MongoDB.Analyzer/MongoDB.Analyzer.csproj (hunk 8 above)
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/MongoDB.Analyzer.csproj:34]: Done

File: src/MongoDB.Analyzer/Core/Utilities/AnalysisCodeGeneratorUtilities.cs (hunk 9 above)
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/Core/Utilities/AnalysisCodeGeneratorUtilities.cs:28]: Done

File: src/MongoDB.Analyzer/Core/Utilities/AnalysisCodeGeneratorUtilities.cs (hunk 6 above)
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/Core/Utilities/AnalysisCodeGeneratorUtilities.cs:63]: Done

File: tests/MongoDB.Analyzer.Tests/Infrastructure/DiagnosticsAnalyzer.cs (hunk 7 above)
Comment (Ravi-Raghavan) [reply tests/MongoDB.Analyzer.Tests/Infrastructure/DiagnosticsAnalyzer.cs:59]: Done

**Review by BorisDog — CHANGES_REQUESTED** [review]

File: src/MongoDB.Analyzer/Core/Utilities/AnalysisCodeGeneratorUtilities.cs [hunk 10]
```diff
@@ -56,5 +60,7 @@ analysisType switch
             AnalysisType.Poco => JsonSyntaxElements.Poco.JsonGeneratorFullName,
             _ => throw new ArgumentOutOfRangeException(nameof(analysisType), analysisType, "Unsupported analysis type")
         };
+
+    private static bool IsDriverVersion_2_28_Or_Greater(Version driverVersion) => driverVersion >= s_driverVersion_2_28;
```
[the remark this answers is not in the document]
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/Core/Utilities/AnalysisCodeGeneratorUtilities.cs:64]: Done

File: tests/MongoDB.Analyzer.Tests/Infrastructure/DiagnosticsAnalyzer.cs [hunk 11]
```diff
@@ -31,7 +31,9 @@ internal static class DiagnosticsAnalyzer
         Common.LinqVersion linqVersion,
         Common.PocoAnalysisVerbosity jsonAnalyzerVerbosity)
     {
-        PathUtilities.VerifyTestDataModelAssembly();
+        var isDriverVersion_2_28_OrGreater = PathUtilities.IsDriverVersion_2_28_OrGreater(driverVersion);
+        var testDataModelAssembly = isDriverVersion_2_28_OrGreater ? PathUtilities.TestDataModelAssemblyPathDriver_2_28_OrGreater : PathUtilities.TestDataModelAssemblyPathDriver_2_27_OrLower;
+        PathUtilities.VerifyTestDataModelAssembly(testDataModelAssembly);
```
[the remark this answers is not in the document]
Comment (Ravi-Raghavan) [reply tests/MongoDB.Analyzer.Tests/Infrastructure/DiagnosticsAnalyzer.cs:36]: Done

File: src/MongoDB.Analyzer/Core/Utilities/AnalysisCodeGeneratorUtilities.cs [hunk 12]
```diff
@@ -18,15 +18,19 @@ namespace MongoDB.Analyzer.Core.Utilities;
...
     public static Type CompileAndGetGeneratorType(AnalysisType analysisType, MongoAnalysisContext context, ReferencesContainer referencesContainer, IEnumerable<SyntaxTree> syntaxTrees)
     {
+        var assemblyName = referencesContainer.Version >= s_driverVersion_2_28 ? AnalysisAssemblyNameDriver_2_28_OrGreater : AnalysisAssemblyNameDriver_2_27_OrLower;
         var compilation = CSharpCompilation.Create(
-            AnalysisAssemblyName,
+            assemblyName,
             syntaxTrees,
             referencesContainer.References,
-            new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary));
+            new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary, publicSign: true, cryptoPublicKey: s_publicKey.ToImmutableArray()));
```
Comment (BorisDog) [inline src/MongoDB.Analyzer/Core/Utilities/AnalysisCodeGeneratorUtilities.cs:33]: make `s_publicKey` immutable array, to skip the conversion each time.

**Review by BorisDog — APPROVED** [review]
LGTM, good job!

File: src/MongoDB.Analyzer/Core/Utilities/AnalysisCodeGeneratorUtilities.cs (hunk 12 above)
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/Core/Utilities/AnalysisCodeGeneratorUtilities.cs]: Done

**Pull request merged**